### PR TITLE
*: use FnvHashMap and FnvHashSet

### LINKED ISSUE #1131: use FnvHashMap instead of origin SipHashMap with integer key.
We have may HashMap with integer key, maybe we can use FnvHashMap instead of origin SipHashMap, because it is more fast, see https://blog.rust-lang.org/2016/03/02/Rust-1.7.html

We should also do some benchmark.

### PR DESCRIPTION
**Pull request by overvenus**
Replace SipHash 1-3 hasher with Fnv hasher. ~~Mod raft remains SipHash 1-3.~~

### Uniqueness and speed?

 + http://softwareengineering.stackexchange.com/questions/49550/which-hashing-algorithm-is-best-for-uniqueness-and-speed
 + https://131002.net/siphash/

PTAL @siddontang @BusyJay 

Close #1131

### CODE REVIEW

**Comment by BusyJay** [comment]
raft mod should be updated too.

**Comment by siddontang** [comment]
any benchmark?

File: src/util/macros.rs [hunk 1]
```diff
@@ -67,13 +67,16 @@ macro_rules! map {
...
         {
             use std::collections::HashMap;
-            HashMap::new()
+            use fnv::FnvBuildHasher;
+            HashMap::with_hasher(FnvBuildHasher::default())
         }
     };
     ( $( $k:expr => $v:expr ),+ ) => {
         {
             use std::collections::HashMap;
```
Comment (ngaut) [inline src/util/macros.rs]: Remove this line.

**Comment by overvenus** [comment]
```
## fnv
benching Set on raft cluster (channel), nodes: 1, value len:    8       ...     58,890 ns/iter (+/- 50,087)
benching Set on raft cluster (channel), nodes: 1, value len:  128       ...     34,617 ns/iter (+/- 15,481)
benching Set on raft cluster (channel), nodes: 1, value len: 1024       ...     43,351 ns/iter (+/- 11,903)
benching Set on raft cluster (channel), nodes: 1, value len: 4096       ...     96,363 ns/iter (+/- 134,625)
benching Get on raft cluster (channel), nodes: 1        ...     20,302 ns/iter (+/- 11,237)
benching Delete on raft cluster (channel),      nodes: 1        ...     50,544 ns/iter (+/- 70,306)
benching Set on raft cluster (channel), nodes: 3, value len:    8       ...    149,454 ns/iter (+/- 198,424)
benching Set on raft cluster (channel), nodes: 3, value len:  128       ...    183,550 ns/iter (+/- 159,142)
benching Set on raft cluster (channel), nodes: 3, value len: 1024       ...     97,274 ns/iter (+/- 21,832)
benching Set on raft cluster (channel), nodes: 3, value len: 4096       ...    201,024 ns/iter (+/- 303,104)
benching Get on raft cluster (channel), nodes: 3        ...     21,170 ns/iter (+/- 18,971)
benching Delete on raft cluster (channel),      nodes: 3        ...    205,488 ns/iter (+/- 191,449)
benching Set on raft cluster (channel), nodes: 5, value len:    8       ...    192,068 ns/iter (+/- 303,518)
benching Set on raft cluster (channel), nodes: 5, value len:  128       ...     97,811 ns/iter (+/- 26,459)
benching Set on raft cluster (channel), nodes: 5, value len: 1024       ...    228,127 ns/iter (+/- 356,491)
benching Set on raft cluster (channel), nodes: 5, value len: 4096       ...    142,074 ns/iter (+/- 80,116)
benching Get on raft cluster (channel), nodes: 5        ...     18,782 ns/iter (+/- 16,140)
benching Delete on raft cluster (channel),      nodes: 5        ...    190,762 ns/iter (+/- 351,812)
benching Set on raft cluster (tcp),     nodes: 1, value len:    8       ...    103,470 ns/iter (+/- 203,572)
benching Set on raft cluster (tcp),     nodes: 1, value len:  128       ...    101,868 ns/iter (+/- 77,276)
benching Set on raft cluster (tcp),     nodes: 1, value len: 1024       ...    129,003 ns/iter (+/- 102,858)
benching Set on raft cluster (tcp),     nodes: 1, value len: 4096       ...    197,963 ns/iter (+/- 290,567)
benching Get on raft cluster (tcp),     nodes: 1        ...     97,719 ns/iter (+/- 135,375)
benching Delete on raft cluster (tcp),  nodes: 1        ...    113,834 ns/iter (+/- 132,562)
benching Set on raft cluster (tcp),     nodes: 3, value len:    8       ...    540,982 ns/iter (+/- 697,501)
benching Set on raft cluster (tcp),     nodes: 3, value len:  128       ...    319,137 ns/iter (+/- 516,492)
benching Set on raft cluster (tcp),     nodes: 3, value len: 1024       ...    515,092 ns/iter (+/- 651,795)
benching Set on raft cluster (tcp),     nodes: 3, value len: 4096       ...    765,158 ns/iter (+/- 844,028)
benching Get on raft cluster (tcp),     nodes: 3        ...     85,752 ns/iter (+/- 142,009)
benching Delete on raft cluster (tcp),  nodes: 3        ...    308,535 ns/iter (+/- 139,837)
benching Set on raft cluster (tcp),     nodes: 5, value len:    8       ...    626,818 ns/iter (+/- 688,391)
benching Set on raft cluster (tcp),     nodes: 5, value len:  128       ...    665,451 ns/iter (+/- 1,110,828)
benching Set on raft cluster (tcp),     nodes: 5, value len: 1024       ...    731,471 ns/iter (+/- 1,064,444)
benching Set on raft cluster (tcp),     nodes: 5, value len: 4096       ...    580,160 ns/iter (+/- 1,065,199)
benching Get on raft cluster (tcp),     nodes: 5        ...     41,589 ns/iter (+/- 12,618)
benching Delete on raft cluster (tcp),  nodes: 5        ...    398,594 ns/iter (+/- 661,861)
benching tombstone scan with rocksdb    ...      47,045,208 ns/iter (+/- 40,309,752)


## master
benching Set on raft cluster (channel), nodes: 1, value len:    8       ...     52,500 ns/iter (+/- 75,585)
benching Set on raft cluster (channel), nodes: 1, value len:  128       ...     45,525 ns/iter (+/- 34,592)
benching Set on raft cluster (channel), nodes: 1, value len: 1024       ...     53,152 ns/iter (+/- 25,248)
benching Set on raft cluster (channel), nodes: 1, value len: 4096       ...     88,665 ns/iter (+/- 108,784)
benching Get on raft cluster (channel), nodes: 1        ...     18,967 ns/iter (+/- 18,239)
benching Delete on raft cluster (channel),      nodes: 1        ...     39,373 ns/iter (+/- 11,766)
benching Set on raft cluster (channel), nodes: 3, value len:    8       ...    192,353 ns/iter (+/- 244,619)
benching Set on raft cluster (channel), nodes: 3, value len:  128       ...    202,555 ns/iter (+/- 197,131)
benching Set on raft cluster (channel), nodes: 3, value len: 1024       ...     95,528 ns/iter (+/- 36,604)
benching Set on raft cluster (channel), nodes: 3, value len: 4096       ...    332,776 ns/iter (+/- 341,392)
benching Get on raft cluster (channel), nodes: 3        ...     21,140 ns/iter (+/- 11,576)
benching Delete on raft cluster (channel),      nodes: 3        ...    199,252 ns/iter (+/- 256,536)
benching Set on raft cluster (channel), nodes: 5, value len:    8       ...    161,976 ns/iter (+/- 273,676)
benching Set on raft cluster (channel), nodes: 5, value len:  128       ...    113,789 ns/iter (+/- 246,665)
benching Set on raft cluster (channel), nodes: 5, value len: 1024       ...    242,130 ns/iter (+/- 289,161)
benching Set on raft cluster (channel), nodes: 5, value len: 4096       ...    293,991 ns/iter (+/- 284,924)
benching Get on raft cluster (channel), nodes: 5        ...     21,296 ns/iter (+/- 18,199)
benching Delete on raft cluster (channel),      nodes: 5        ...    220,176 ns/iter (+/- 270,121)
benching Set on raft cluster (tcp),     nodes: 1, value len:    8       ...    122,951 ns/iter (+/- 157,153)
benching Set on raft cluster (tcp),     nodes: 1, value len:  128       ...    126,305 ns/iter (+/- 172,622)
benching Set on raft cluster (tcp),     nodes: 1, value len: 1024       ...    171,781 ns/iter (+/- 264,091)
benching Set on raft cluster (tcp),     nodes: 1, value len: 4096       ...    159,411 ns/iter (+/- 229,754)
benching Get on raft cluster (tcp),     nodes: 1        ...    101,870 ns/iter (+/- 129,397)
benching Delete on raft cluster (tcp),  nodes: 1        ...     96,515 ns/iter (+/- 189,920)
benching Set on raft cluster (tcp),     nodes: 3, value len:    8       ...    324,724 ns/iter (+/- 509,918)
benching Set on raft cluster (tcp),     nodes: 3, value len:  128       ...    548,293 ns/iter (+/- 660,251)
benching Set on raft cluster (tcp),     nodes: 3, value len: 1024       ...    738,373 ns/iter (+/- 924,848)
benching Set on raft cluster (tcp),     nodes: 3, value len: 4096       ...    411,412 ns/iter (+/- 415,668)
benching Get on raft cluster (tcp),     nodes: 3        ...     90,043 ns/iter (+/- 153,418)
benching Delete on raft cluster (tcp),  nodes: 3        ...    634,147 ns/iter (+/- 727,943)
benching Set on raft cluster (tcp),     nodes: 5, value len:    8       ...    737,243 ns/iter (+/- 823,362)
benching Set on raft cluster (tcp),     nodes: 5, value len:  128       ...    355,984 ns/iter (+/- 160,497)
benching Set on raft cluster (tcp),     nodes: 5, value len: 1024       ...    594,521 ns/iter (+/- 771,352)
benching Set on raft cluster (tcp),     nodes: 5, value len: 4096       ...    451,718 ns/iter (+/- 183,050)
benching Get on raft cluster (tcp),     nodes: 5        ...     90,292 ns/iter (+/- 128,373)
benching Delete on raft cluster (tcp),  nodes: 5        ...    685,775 ns/iter (+/- 872,542)
benching tombstone scan with rocksdb    ...      52,046,728 ns/iter (+/- 47,216,540)
```

**Comment by ngaut** [comment]
Good, Faster!

**Comment by siddontang** [comment]
> Mod raft remains SipHash 1-3.

Why still use SipHash for raft?

**Comment by overvenus** [comment]
@siddontang  Oh, sorry, I forgot to update description. Now, fully switch to fnv hasher. Although I still concern about collision.

**Comment by siddontang** [comment]
LGTM

Is it easy for us to change using Sip again?

**Comment by siddontang** [comment]
Please fix the conflict. @overvenus

**Comment by ngaut** [comment]
LGTM

**Comment by overvenus** [comment]
@siddontang Sure, it is really simple to change to another hasher, as long as the hasher `impl`s trait `Default`.

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -39,8 +40,8 @@ use std::path::Path;
 use std::sync::Arc;
 use std::io::Read;
 use std::time::Duration;
-use std::collections::HashMap;
 
+use fnv::FnvHashMap as HashMap;
```
Comment (overvenus) [inline src/bin/tikv-server.rs]: Oops, I forgot it. It should be `tikv::util::HashMap`.

**Pull request merged**